### PR TITLE
Add support for Razer Lancehead

### LINKED ISSUE #540: Support for Lancehead wireless mouse?
Greetings,
 
I installed openrazer tonight and noticed it does not appear to support the Lancehead wireless mouse.  Is there anything I could do to help create support for this mouse?  

I have a functioning install of the driver -- it notices my Naga Hex V2 just fine. 

My Lancehead has model number: RC30-021203

Product No: RZ01-02120100

After an update Synapse is recognising the mouse now.  I'll upload pictures of the Synapse pages for the mouse.

### PR DESCRIPTION
**Pull request by nightsky30**
Based on what was done for the Lancehead TE, I've added the pieces required for the Lancehead (0x0059, 0x005a).  From initial testing by a few helpful people, everything seems to be on par with the Lancehead TE regarding detection, dpi, poll rate, brightness, and lighting effects.  This fixes #540.

My only real concern is the battery/charging status support.  I've added what I think is necessary.~~, though I wasn't sure how the battery/charging status is normally shown.  Is it just by screen notifications?~~

~~Definitely a work in progress that could use more testing, but I think it's 90% there.~~

**Tested and works:**
 - [x] Wired
 - [x] Wireless
 - [x] Battery Status

### CODE REVIEW

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 1]
```diff
@@ -15,6 +15,156 @@
     set_left_brightness as _set_left_brightness, set_right_brightness as _set_right_brightness
 
 
+class RazerLanceheadWireless(__RazerDeviceSpecialBrightnessSuspend):
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: As the two devices are actually one and should support exactly the same things, you can e.g. make the Wireless version inherit from the Wired one and just overwrite the PID and potential other relevant parts.

File: driver/razermouse_driver.c [hunk 2]
```diff
@@ -2655,6 +2765,13 @@ static int razer_mouse_probe(struct hid_device *hdev, const struct hid_device_id
         CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_device_mode);
 
         switch(dev->usb_pid) {
+        case USB_DEVICE_ID_RAZER_LANCEHEAD_WIRELESS:
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_charge_effect);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_charge_colour);
+        case USB_DEVICE_ID_RAZER_LANCEHEAD_WIRED:
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_charge_level);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_charge_status);
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_charge_low_threshold);
```
Comment (z3ntu) [inline driver/razermouse_driver.c:3109]: please add fallthrough comments there so the compiler is definitely happy? With the latest kernel version or so, these warnings are enabled by default in the kernel compilation.

File: driver/razermouse_driver.c [hunk 3]
```diff
@@ -2971,6 +3088,13 @@ static void razer_mouse_disconnect(struct hid_device *hdev)
         device_remove_file(&hdev->dev, &dev_attr_device_mode);
 
         switch(usb_dev->descriptor.idProduct) {
+        case USB_DEVICE_ID_RAZER_LANCEHEAD_WIRELESS:
+            device_remove_file(&hdev->dev, &dev_attr_charge_effect);
+            device_remove_file(&hdev->dev, &dev_attr_charge_colour);
+        case USB_DEVICE_ID_RAZER_LANCEHEAD_WIRED:
+            device_remove_file(&hdev->dev, &dev_attr_charge_level);
+            device_remove_file(&hdev->dev, &dev_attr_charge_status);
+            device_remove_file(&hdev->dev, &dev_attr_charge_low_threshold);
```
Comment (z3ntu) [inline driver/razermouse_driver.c:3535]: same

File: pylib/openrazer/client/fx.py [hunk 4]
```diff
@@ -717,6 +717,16 @@ def static(self, red: int, green: int, blue: int) -> bool:
             return True
         return False
 
+    def wave(self, direction: int) -> bool:
```
Comment (z3ntu) [inline pylib/openrazer/client/fx.py]: Please make the pylib changes a separate PR

**Review by z3ntu — CHANGES_REQUESTED** [review]
See comments, LGTM otherwise

File: daemon/openrazer_daemon/hardware/mouse.py (hunk 1 above)
Comment (nightsky30) [reply daemon/openrazer_daemon/hardware/mouse.py]: Should be good now.

File: driver/razermouse_driver.c (hunk 2 above)
Comment (nightsky30) [reply driver/razermouse_driver.c:3109]: So when trying [[fallthrough]]; as provided by C++17 my system wasn't having it at all.  Not sure why.  I found an article on GCC 7 and various fall-through statements.  The following compiled for me, and seemed like it should work with older compilers as well:

`__attribute__ ((fallthrough)); // C and C++03`

Do you feel this is sufficient, or do you prefer another method (C++11, C++14, etc)?

**For reference:**
https://developers.redhat.com/blog/2017/03/10/wimplicit-fallthrough-in-gcc-7/

https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-fallthrough

**Edit:**  Compiles on my system, but Travis does not seem to like it all that much...I'm using GCC 8.

Found mention elsewhere that anything in kernel space must adhere to C90?  If that is the case, I guess C99 is out of the realm of possible solutions.

I can try C11, C14, C17, or the comment styles again...

File: driver/razermouse_driver.c (hunk 3 above)
Comment (nightsky30) [reply driver/razermouse_driver.c:3535]: See above.

File: driver/razermouse_driver.c (hunk 2 above)
Comment (nightsky30) [reply driver/razermouse_driver.c]: Travis does not seem to like it all that much...

File: pylib/openrazer/client/fx.py (hunk 4 above)
Comment (nightsky30) [reply pylib/openrazer/client/fx.py]: Reverted the changes.  I submitted a separate PR #893.

File: driver/razermouse_driver.c (hunk 2 above)
Comment (nightsky30) [reply driver/razermouse_driver.c:3109]: Went with a comment vs an attribute.

**Comment by nightsky30** [comment]
@z3ntu I've completed the requested changes.  I also added a missing device file attribute for idle time.  Requesting user to retest.

Also tried looking into what's going on with 897, but didn't see anything that jumped out.  Other wireless devices like the Mamba Chroma don't seem to explicitly list get_battery device attribute files in the driver, or am I wrong?

File: driver/razermouse_driver.c [hunk 5]
```diff
@@ -2971,6 +3092,16 @@ static void razer_mouse_disconnect(struct hid_device *hdev)
...
+        case USB_DEVICE_ID_RAZER_LANCEHEAD_WIRELESS:
+            device_remove_file(&hdev->dev, &dev_attr_charge_effect);
+            device_remove_file(&hdev->dev, &dev_attr_charge_colour);
+        /* fallthrough */
+        case USB_DEVICE_ID_RAZER_LANCEHEAD_WIRED:
+            device_remove_file(&hdev->dev, &dev_attr_charge_level);
+            device_remove_file(&hdev->dev, &dev_attr_charge_status);
+            device_remove_file(&hdev->dev, &dev_attr_charge_low_threshold);
+            device_remove_file(&hdev->dev, &dev_attr_device_idle_time);
+        /* fallthrough */
```
Comment (z3ntu) [inline driver/razermouse_driver.c]: indent the comments one more to the right please, to be in line with the statements before it

Comment (nightsky30) [reply driver/razermouse_driver.c]: I made the changes as requested, but then Travis CI complains and the checks fail because this seems to go against the astyle linux formatting.  Running astyle again will revert the changes you requested.

**Comment by jeaye** [comment]
Is it just whitespace changes keeping this guy from being merged? Would love to see it in master. Thanks for the awesome work, @nightsky30!

**Comment by nightsky30** [comment]
> Is it just whitespace changes keeping this guy from being merged? Would love to see it in master. Thanks for the awesome work, @nightsky30!

Battery/Charging status is the only unknown that could use some more testing if users are concerned about it.  Everything else *should* be working based on testing that was done in #540.

**Review by Cybiko — APPROVED** [review]

**Comment by z3ntu** [comment]
@nightsky30 Any updates here (from looking at the last comment there's still some stuff to fix)? Also you can fix the CI by running `./scripts/format_source.sh`

**Comment by nightsky30** [comment]
> @nightsky30 Any updates here (from looking at the last comment there's still some stuff to fix)? Also you can fix the CI by running `./scripts/format_source.sh`

@z3ntu If I remember right, I think I based the battery status stuff off of the Mamba Chroma/2012.  I haven't heard from anyone with a Lancehead on the battery status.  It might be fine?  But from some other issues/comments I've seen in passing it seems like the battery status reporting hasn't been great. #576 #957 Everything is confirmed working with the Lancehead otherwise.

With regard to the CI, I just wanted to confirm you are ok with the /* fallthrough */ comments being indented differently.  If I run the styling script it will undo the change you requested in the review.  It seems the styling format wants the comments to stick out that way.

**Comment by z3ntu** [comment]
Maybe try `// fallthrough` , does that get formatted differently?

**Comment by nightsky30** [comment]
> Maybe try `// fallthrough` , does that get formatted differently?

Unfortunately that sticks out as well :(

**Comment by z3ntu** [comment]
It's fine then.

So, ready to merge?

**Comment by nightsky30** [comment]
Yep, I think it's good.  Thank you z3ntu :)

File: daemon/openrazer_daemon/hardware/mouse.py [hunk 6]
```diff
@@ -15,6 +15,103 @@
...
+        _set_right_brightness(self, right_row_brightness)
+        self.disable_notify = False
+
+
+class RazerLanceheadWireless(RazerLanceheadWired):
+    """
+    Class for the Razer Lancehead (Wireless)
+    """
+    USB_PID = 0x005A
+    METHODS = ['set_charge_effect', 'set_charge_colour']
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse.py]: That looks like it overwrites the other `METHODS` from the `RazerLanceheadWired` class and as such only has `set_charge_effect` & `set_charge_colour` effects and nothing else. 

Was the wireless version actually tested? The rest of this PR looks fine.

Comment (nightsky30) [reply daemon/openrazer_daemon/hardware/mouse.py]: Sorry about that.  Good catch.  I've modified it so that it appends using the + operator.  It was either that or using .extend().  Thought this was simplest.

The wireless testing was conducted with djmulder in #540 at some point, probably before d778072b029ab0414dba3d18d535f1d9fb07d3c7.  I can request a retest and hopefully have someone report back here.

**Review by z3ntu — APPROVED** [review]
LGTM, but I'll wait for somebody to test it before merging as written in #540

**Comment by crian** [comment]
@nightsky30 have you got the battery status working on Lancehead? I'm adding support for the Mamba Wireless and that is one thing that's missing.

**Comment by nightsky30** [comment]
> @nightsky30 have you got the battery status working on Lancehead? I'm adding support for the Mamba Wireless and that is one thing that's missing.

@crian I used the Mamba Chroma/2012 as examples, so it may or may not.  I don't own the Lancehead, but the Lancehead TE.  So it is really dependent on the users to test and let us know.

Edit:

I had added switch cases to **razer_attr_read_get_battery** and **razer_attr_read_is_charging** to set the transaction ID of the report to 0x3f.  I did this because it is required elsewhere (e.g. firmware, serial, poll rate, device mode) for the Lancehead.  If the Mamba battery/charging status isn't working for you currently, then you could try that.  I'd be interested in the outcome on your Mamba **if** it uses the same transaction ID.

**Comment by crian** [comment]
@nightsky30 sorry for the late report. the new switch cases didn't work for the Mamba Wireless (2018).

**Comment by nightsky30** [comment]
No worries!  Thanks for the update. :smile: Maybe not working with the Lancehead either then.

**Comment by nightsky30** [comment]
With the Lancehead Wireless (2019) #1055 and Mamba Wireless #1075 requiring the high `RAZER_MOUSE_WAIT_MIN_US` and `RAZER_MOUSE_WAIT_MAX_US` values along with initial reports from testing the battery on this Lancehead (2017) being unsuccessful, it may be that this mouse requires the higher values as well.

Verifying this possibility is still dependent on the users that own the actual hardware providing the test support.

**Comment by WhyNotHugo** [comment]
My Lancehead wireless reports itself as `1532:0070`. I tried patching this PR to use that ID instead of `005a`, but that just resulted in the mouse not working at all (lights turn on, but it doesn't move).

When plugging the mouse, this is all that the logs have to say:

```
Apr 16 02:08:24 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:24 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.0/0003:1532:0070.000F
Apr 16 02:08:24 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:24 | razer                          | DEBUG    | Device event [bind]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.0/0003:1532:0070.000F
Apr 16 02:08:24 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:24 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.2/0003:1532:0070.0011
Apr 16 02:08:24 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:24 | razer                          | DEBUG    | Device event [bind]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.2/0003:1532:0070.0011
Apr 16 02:08:25 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:25 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.1/0003:1532:0070.0010
Apr 16 02:08:25 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:25 | razer                          | DEBUG    | Device event [bind]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.1/0003:1532:0070.0010
Apr 16 02:08:25 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:25 | razer                          | DEBUG    | Device event [unbind]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.0/0003:1532:0070.000F
Apr 16 02:08:25 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:25 | razer                          | DEBUG    | Device event [unbind]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.2/0003:1532:0070.0011
Apr 16 02:08:25 aphrodite openrazer-daemon[686]: 2020-04-16 02:08:25 | razer                          | DEBUG    | Device event [unbind]: /devices/pci0000:00/0000:00:14.0/usb1/1-4/1-4:1.1/0003:1532:0070.0010
```

Note that merely removing openrazer and unplugging/replugging the mouse made it work again.

**Comment by WhyNotHugo** [comment]
Oh, upon further inspection looks like mine is called "Razer Lancehead Wireless (2019)", and I should look at #1055.

Oddly, I got this in late 2018. Is the naming an upstream thing?

**Comment by nightsky30** [comment]
The official names are "Lancehead" released in Apr 2017, and "Lancehead Wireless" released in May 2019.  I just listed the (2019) here to help differentiate the two.  But yes, looks like you need #1055

**Comment by z3ntu** [comment]
@nightsky30 This PR still hasn't been tested on an actual device, right? So we're still waiting for that

**Comment by nightsky30** [comment]
@z3ntu The best we have is it was tested ~2 weeks ago in https://github.com/openrazer/openrazer/issues/540#issuecomment-612464172.  @amit-feldman did confirm the device works wirelessly, but I could not get a confirmation on the battery notifications.

**Comment by amit-feldman** [comment]
Sorry for the late reply, @nightsky30. I did get correct wireless battery notifications.

**Comment by nightsky30** [comment]
@z3ntu Looks good to go!  Thanks again for testing @amit-feldman :smile:

**Review by z3ntu — APPROVED** [review]
LGTM, thanks!
